src/katsdpmodels/fetch.py

```diff
@@ -0,0 +1,359 @@
+        If true, the session used for HTTP requests will be closed when the
+        fetcher is closed. It defaults to true if an internally-created session
+        is used, and false if the user provided a session.
+
+    Raises
+    ------
+    :exc:`.ModelError`
+        For any issues with the model itself
+    :exc:`requests.exceptions.RequestException`
+        For any issues at the HTTP level
+    """
+
+    def __init__(self, *, session: Optional[Session] = None) -> None:
+        self._session: Session
+        if session is None:
+            self._session = requests.Session()
+            # TODO: requests_file is convenient, but it would be more efficient to
+            # open the file directly with h5py rather than sucking it into a
+            # BytesIO.
+            self._session.mount('file://', requests_file.FileAdapter())
+            self.close_session = True
+        else:
+            self._session = session
+            self.close_session = False
+        self._alias_cache: Dict[str, str] = {}
+        self._model_cache: Dict[str, models.Model] = {}
+        self.close_models = True
+
+    @property
+    def session(self) -> Session:
+        return self._session
+
+    def close(self) -> None:
+        """Release the resources associated with the fetcher.
+
+        See also
+        --------
+        :attr:`close_session`, :attr:`close_models`
+        """
+        if self.close_session:
+            self._session.close()
+        self._alias_cache.clear()
+        if self.close_models:
+            for model in self._model_cache.values():
+                model.close()
+        self._model_cache.clear()
+
+    def resolve(self, url: str) -> str:
+        """Find the canonical URL, following aliases."""
+        original_url = url
+        aliases = 0
+        parts = urllib.parse.urlparse(url)
+        while parts.path.endswith('.alias'):
+            aliases += 1
+            if aliases > MAX_ALIASES:
+                raise models.TooManyAliasesError.with_urls(
+                    f'Reached limit of {MAX_ALIASES} levels of aliases',
+                    url=url, original_url=original_url)
+            if url in self._alias_cache:
+                new_url = self._alias_cache[url]
+            else:
+                with self._session.get(url) as resp:
+                    resp.raise_for_status()
+                    rel_path = resp.text.rstrip()
+                    new_url = urllib.parse.urljoin(resp.url, rel_path)
+                self._alias_cache[url] = new_url
+            _logger.debug('Redirecting from %s to %s', url, new_url)
+            url = new_url
+            parts = urllib.parse.urlparse(url)
+        return url
+
+    def _get_eager(self, url: str) -> Tuple[io.IOBase, str, str]:
+        with self._session.get(url) as resp:
+            resp.raise_for_status()
+            data = resp.content
+            url = resp.url     # Handle HTTP redirects
+        checksum = hashlib.sha256(data).hexdigest()
```

Comment:
I think I recall MD5 being quicker than SHA256 - any reason for not using MD5?

Reply:
MD5 is broken. Not that a collision attack seems like something we really need to worry about, but I just avoid it on principal. A quick test on my laptop shows that SHA256 does about 500MB/s, which is less than I'd hoped but I don't think is a train smash.

Reply:
I was concerned about large models here. As you say, MD5 is vulnerable, but I really don't think we are trying to prevent purposeful attacks with the checksum - rather detect errors. If you are happy with the performance then fine.

Reply:
I'm happy. And it seems to be implementation/machine-dependent; according to [this benchmark](http://bench.cr.yp.to/results-hash.html#amd64-genji346) SHA256 can be 2.5x faster than MD5 in some cases.